README.md
# tedious-connection-pool
[![Dependency Status](https://david-dm.org/tediousjs/tedious-connection-pool.svg)](https://david-dm.org/tediousjs/tedious-connection-pool)
[![npm version](https://badge.fury.io/js/tedious-connection-pool.svg)](https://badge.fury.io/js/tedious-connection-pool)
[![Build status](https://ci.appveyor.com/api/projects/status/jnurb48ao1wrbgbr?svg=true)](https://ci.appveyor.com/project/ben-page/tedious-connection-pool)


A connection pool for [tedious](http://github.com/tediousjs/tedious).

## Installation

    npm install tedious-connection-pool
    
## Description
The only difference from the regular tedious API is how the connection is obtained and released. Rather than creating a connection and then closing it when finished, acquire a connection from the pool and release it when finished. Releasing resets the connection and makes in available for another use.

Once the Tedious Connection object has been acquired, the tedious API can be used with the connection as normal.

## Example

```javascript
var ConnectionPool = require('tedious-connection-pool');
var Request = require('tedious').Request;

var poolConfig = {
    min: 2,
    max: 4,
    log: true
};

var connectionConfig = {
    userName: 'login',
    password: 'password',
    server: 'localhost'
};

//create the pool
var pool = new ConnectionPool(poolConfig, connectionConfig);

pool.on('error', function(err) {
    console.error(err);
});

//acquire a connection
pool.acquire(function (err, connection) {
    if (err) {
        console.error(err);
        return;
    }

    //use the connection as normal
    var request = new Request('select 42', function(err, rowCount) {
        if (err) {
            console.error(err);
            return;
        }

        console.log('rowCount: ' + rowCount);

        //release the connection back to the pool when finished
        connection.release();
    });

    request.on('row', function(columns) {
        console.log('value: ' + columns[0].value);
    });

    connection.execSql(request);
});
```

When you are finished with the pool, you can drain it (close all connections).
```javascript
pool.drain();
```


## Class: ConnectionPool

### new ConnectionPool(poolConfig, connectionConfig)

* `poolConfig` {Object} the pool configuration object
  * `min` {Number} The minimun of connections there can be in the pool. Default = `10`
  * `max` {Number} The maximum number of connections there can be in the pool. Default = `50`
  * `idleTimeout` {Number} The number of milliseconds before closing an unused connection. Default = `300000`
  * `retryDelay` {Number} The number of milliseconds to wait after a connection fails, before trying again. Default = `5000`
  * `acquireTimeout` {Number} The number of milliseconds to wait for a connection, before returning an error. Default = `60000`
  * `log` {Boolean|Function} Set to true to have debug log written to the console or pass a function to receive the log messages. Default = `undefined`
  
* `connectionConfig` {Object} The same configuration that would be used to [create a
  tedious Connection](http://pekim.github.com/tedious/api-connection.html#function_newConnection).

### connectionPool.acquire(callback)
Acquire a Tedious Connection object from the pool.

 * `callback(err, connection)` {Function} Callback function
  * `err` {Object} An Error object is an error occurred trying to acquire a connection, otherwise null.
  * `connection` {Object} A [Connection](http://pekim.github.com/tedious/api-connection.html)

### connectionPool.drain(callback)
Close all pooled connections and stop making new ones. The pool should be discarded after it has been drained.
 * `callback()` {Function} Callback function

### connectionPool.error {event}
The 'error' event is emitted when a connection fails to connect to the SQL Server. The pool will simply retry indefinitely. The application may want to handle errors in a more nuanced way.

## Class: Connection
The following method is added to the Tedious [Connection](http://pekim.github.com/tedious/api-connection.html) object.

### Connection.release()
Release the connect back to the pool to be used again

## Changelog

### Version 1.0.3
* bug fix only

### Version 1.0.3
* Pool modifies the Tedious connection object rather than the Connection prototype.

### Version 1.0.2
* Added additional log message when acquiring a connection.

### Version 1.0.0
* No changes from v0.3.9.

### Version 0.3.9
* bug fix only

### Version 0.3.7
* bug fix only

### Version 0.3.6
* bug fix only

### Version 0.3.5
* `poolConfig` option `min` is limited to less than `max`

### Version 0.3.4
* `poolConfig` option `min` supports being set to 0

### Version 0.3.3
* Ignore calls to connection.release() on a connection that has been closed or not part of the connection pool.

### Version 0.3.2
 * Calls connection.reset() when the connection is released to the pool. This is very unlikely to cause anyone trouble.
 * Added a callback argument to connectionPool.drain()

### Version 0.3.0
 * Removed dependency on the `generic-pool` node module.
 * Added `poolConfig` options `retryDelay`
 * Added `poolConfig` options `aquireTimeout` **(Possibly Breaking)**
 * Added `poolConfig` options `log`
 * `idleTimeoutMillis` renamed to `idleTimeout` **(Possibly Breaking)**
 * The `ConnectionPool` `'error'` event added
 * The behavior of the err parameter of the callback passed to `acquire()` has changed. It only returns errors related to acquiring a connection not Tedious Connection errors. Connection errors can happen anytime the pool is being filled and could go unnoticed if only passed the the callback. Subscribe to the `'error'` event on the pool to be notified of all connection errors. **(Possibly Breaking)**
 * `PooledConnection` object removed.

### Version 0.2.x
* To acquire a connection, call on `acquire()` on a `ConnectionPool` rather than `requestConnection()`. **(Breaking)**
* After acquiring a `PooledConnection`, do not wait for the `'connected'` event. The connection is received connected. **(Breaking)**
* Call `release()` on a `PooledConnection` to release the it back to the pool. `close()` permanently closes the connection (as `close()` behaves in in tedious). **(Breaking)**


CHANGELOG.md
### Version 1.0.4
* bug fix only

### Version 1.0.3
* Pool modifies the Tedious connection object rather than the Connection prototype.

### Version 1.0.2
* Added additional log message when acquiring a connection.

### Version 1.0.0
* No changes from v0.3.9.

### Version 0.3.9
* bug fix only

### Version 0.3.7
* bug fix only

### Version 0.3.6
* bug fix only

### Version 0.3.5
* `poolConfig` option `min` is limited to less than `max`

### Version 0.3.4
* `poolConfig` option `min` supports being set to 0

### Version 0.3.3
* Ignore calls to connection.release() on a connection that has been closed or not part of the connection pool.

### Version 0.3.2
 * Calls connection.reset() when the connection is released to the pool. This is very unlikely to cause anyone trouble.
 * Added a callback argument to connectionPool.drain()

### Version 0.3.0
 * Removed dependency on the `generic-pool` node module.
 * Added `poolConfig` options `retryDelay`
 * Added `poolConfig` options `aquireTimeout` **(Possibly Breaking)**
 * Added `poolConfig` options `log`
 * `idleTimeoutMillis` renamed to `idleTimeout` **(Possibly Breaking)**
 * The `ConnectionPool` `'error'` event added
 * The behavior of the err parameter of the callback passed to `acquire()` has changed. It only returns errors related to acquiring a connection not Tedious Connection errors. Connection errors can happen anytime the pool is being filled and could go unnoticed if only passed the the callback. Subscribe to the `'error'` event on the pool to be notified of all connection errors. **(Possibly Breaking)**
 * `PooledConnection` object removed.

### Version 0.2.x
* To acquire a connection, call on `acquire()` on a `ConnectionPool` rather than `requestConnection()`. **(Breaking)**
* After acquiring a `PooledConnection`, do not wait for the `'connected'` event. The connection is received connected. **(Breaking)**
* Call `release()` on a `PooledConnection` to release the it back to the pool. `close()` permanently closes the connection (as `close()` behaves in in tedious). **(Breaking)**


test/test.js
'use strict';
var assert = require('assert');
var Request = require('tedious').Request;
var ConnectionPool = require('../lib/connection-pool');
var Connection = require('tedious').Connection;

var connectionConfig = {
    userName: 'test',
    password: 'test',
    server: 'dev1',
    options: {
        appName: 'pool-test',
        database: 'test'
    }
};

/* create a db user with the correct permissions:
CREATE DATABASE test
CREATE LOGIN test WITH PASSWORD=N'test', DEFAULT_DATABASE=test, CHECK_POLICY=OFF
GRANT ALTER ANY CONNECTION TO test

USE test
CREATE USER test FOR LOGIN test WITH DEFAULT_SCHEMA=dbo
ALTER ROLE db_owner ADD MEMBER test

USE msdb
CREATE USER test FOR LOGIN test WITH DEFAULT_SCHEMA=dbo
ALTER ROLE SQLAgentOperatorRole ADD MEMBER test
ALTER ROLE SQLAgentReaderRole ADD MEMBER test
ALTER ROLE SQLAgentUserRole ADD MEMBER test
*/

/* disable the user when not testing:
ALTER LOGIN test DISABLE
*/

describe('Name Collision', function () {

    it('release', function () {
        assert(!Connection.prototype.release);

        var con = new Connection({});
        assert(!con.release);
        con.close();
    });

    it('pool', function () {
        assert(!Connection.prototype.pool);

        var con = new Connection({});
        assert(!con.pool);
        con.close();
    });
});

describe('ConnectionPool', function () {

    it('min', function (done) {
        this.timeout(10000);

        var poolConfig = {min: 2};
        var pool = new ConnectionPool(poolConfig, connectionConfig);

        setTimeout(function() {
            assert.equal(pool.connections.length, poolConfig.min);
            pool.drain(done);
        }, 4);
    });

    it('min=0', function (done) {
        this.timeout(10000);

        var poolConfig = {min: 0, idleTimeout: 10};
        var pool = new ConnectionPool(poolConfig, connectionConfig);

        setTimeout(function() {
            assert.equal(pool.connections.length, 0);
        }, 4);

        setTimeout(function() {
            pool.acquire(function(err, connection) {
                assert(!err);

                var request = new Request('select 42', function (err, rowCount) {
                    assert.strictEqual(rowCount, 1);
                    connection.release();
                    setTimeout(function () {
                        assert.equal(pool.connections.length, 0);
                        pool.drain(done);
                    }, 200);
                });

                request.on('row', function (columns) {
                    assert.strictEqual(columns[0].value, 42);
                });

                connection.execSql(request);
            });
        }, 2000);
    });

    it('max', function (done) {
        this.timeout(10000);

        var poolConfig = {min: 2, max: 5};
        var pool = new ConnectionPool(poolConfig, connectionConfig);

        var count = 20;
        var run = 0;

        var createRequest = function (err, connection) {
            assert(!err);

            var request = new Request('select 42', function (err, rowCount) {
                assert.strictEqual(rowCount, 1);
                setTimeout(function() {
                    run++;
                    assert(pool.connections.length <= poolConfig.max);
                    if (run === count) {
                        pool.drain(done);
                        return;
                    }
                    connection.release();
                }, 200);
            });

            request.on('row', function (columns) {
                assert.strictEqual(columns[0].value, 42);
            });

            connection.execSql(request);
        };

        for (var i = 0; i < count; i++) {
            setTimeout(function() {
                pool.acquire(createRequest);
            }, 1);
        }
    });

    it('min<=max, min specified > max specified', function (done) {
        this.timeout(10000);

        var poolConfig = { min: 5, max: 1, idleTimeout: 10};
        var pool = new ConnectionPool(poolConfig, connectionConfig);

        setTimeout(function() {
            assert.equal(pool.connections.length, 1);
        }, 4);

        setTimeout(function() {
            pool.acquire(function(err, connection) {
                assert(!err);

                var request = new Request('select 42', function (err, rowCount) {
                    assert.strictEqual(rowCount, 1);
                    connection.release();
                    setTimeout(function () {
                        assert.equal(pool.connections.length, 1);
                        pool.drain(done);
                    }, 200);
                });

                request.on('row', function (columns) {
                    assert.strictEqual(columns[0].value, 42);
                });

                connection.execSql(request);
            });
        }, 2000);
    });

    it('min<=max, no min specified', function (done) {
        this.timeout(10000);

        var poolConfig = {max: 1, idleTimeout: 10};
        var pool = new ConnectionPool(poolConfig, connectionConfig);

        setTimeout(function() {
            assert.equal(pool.connections.length, 1);
        }, 4);

        setTimeout(function() {
            pool.acquire(function(err, connection) {
                assert(!err);

                var request = new Request('select 42', function (err, rowCount) {
                    assert.strictEqual(rowCount, 1);
                    connection.release();
                    setTimeout(function () {
                        assert.equal(pool.connections.length, 1);
                        pool.drain(done);
                    }, 200);
                });

                request.on('row', function (columns) {
                    assert.strictEqual(columns[0].value, 42);
                });

                connection.execSql(request);
            });
        }, 2000);
    });

    it('pool error event', function (done) {
        this.timeout(10000);
        var poolConfig = {min: 3};
        var pool = new ConnectionPool(poolConfig, {});

        pool.acquire(function() { });

        pool.on('error', function(err) {
            assert(!!err);
            pool.drain(done);
        });
    });

    it('connection retry', function (done) {
        this.timeout(10000);
        var poolConfig = {min: 1, max: 5, retryDelay: 5};
        var pool = new ConnectionPool(poolConfig, {});

        pool.on('error', function(err) {
            assert(!!err);
            pool.connectionConfig = connectionConfig;
        });

        function testConnected() {
            for (var i = pool.connections.length - 1; i >= 0; i--) {
                if (pool.connections[i].status === 3/*RETRY*/) {
                    setTimeout(testConnected, 100);
                    return;
                }
            }

            assert.equal(pool.connections.length, poolConfig.min);
            pool.drain(done);
        }

        setTimeout(testConnected, 100);
    });

    it('acquire timeout', function (done) {
        this.timeout(10000);

        var poolConfig = {min: 1, max: 1, acquireTimeout: 2000};
        var pool = new ConnectionPool(poolConfig, connectionConfig);

        pool.acquire(function(err, connection) {
            assert(!err);
            assert(!!connection);
        });

        pool.acquire(function(err, connection) {
            assert(!!err);
            assert(!connection);
            done();
        });
    });

    it('idle timeout', function (done) {
        this.timeout(10000);
        var poolConfig = {min: 1, max: 5, idleTimeout: 100};
        var pool = new ConnectionPool(poolConfig, connectionConfig);

        setTimeout(function() {
            pool.acquire(function (err, connection) {
                assert(!err);

                var request = new Request('select 42', function (err, rowCount) {
                    assert.strictEqual(rowCount, 1);
                    pool.drain(done);
                });

                request.on('row', function (columns) {
                    assert.strictEqual(columns[0].value, 42);
                });

                connection.execSql(request);
            });

        }, 300);
    });

    it('connection error handling', function (done) {
        this.timeout(10000);
        var poolConfig = {min: 1, max: 5};

        var pool = new ConnectionPool(poolConfig, connectionConfig);

        pool.on('error', function(err) {
            assert(err && err.name === 'ConnectionError');
        });

        //This simulates a lost connections by creating a job that kills the current session and then deletesthe job.
        pool.acquire(function (err, connection) {
            assert(!err);

            var command = 'DECLARE @jobName VARCHAR(68) = \'pool\' + CONVERT(VARCHAR(64),NEWID()), @jobId UNIQUEIDENTIFIER;' +
            'EXECUTE msdb..sp_add_job @jobName, @owner_login_name=\'' + connectionConfig.userName + '\', @job_id=@jobId OUTPUT;' +
            'EXECUTE msdb..sp_add_jobserver @job_id=@jobId;' +

            'DECLARE @cmd VARCHAR(50);' +
            'SELECT @cmd = \'kill \' + CONVERT(VARCHAR(10), @@SPID);' +
            'EXECUTE msdb..sp_add_jobstep @job_id=@jobId, @step_name=\'Step1\', @command = @cmd, @database_name = \'' + connectionConfig.options.database + '\', @on_success_action = 3;' +

            'DECLARE @deleteCommand VARCHAR(200);' +
            'SET @deleteCommand = \'execute msdb..sp_delete_job @job_name=\'\'\'+@jobName+\'\'\'\';' +
            'EXECUTE msdb..sp_add_jobstep @job_id=@jobId, @step_name=\'Step2\', @command = @deletecommand;' +

            'EXECUTE msdb..sp_start_job @job_id=@jobId;' +
            'WAITFOR DELAY \'01:00:00\';' +
            'SELECT 42';

            var request = new Request(command, function (err, rowCount) {
                assert(err);
                pool.drain(done);
            });

            request.on('row', function (columns) {
                assert(false);
            });

            connection.execSql(request);
        });
    });

    it('release(), reset()', function (done) {
        this.timeout(10000);

        var poolConfig = {max: 1};
        var pool = new ConnectionPool(poolConfig, connectionConfig);

        var createRequest = function(query, value, callback) {
            var request = new Request(query, function (err, rowCount) {
                assert.strictEqual(rowCount, 1);
                callback();
            });

            request.on('row', function (columns) {
                assert.strictEqual(columns[0].value, value);
            });

            return request;
        };

        pool.acquire(function(err, conn) {
            assert(!err);

            conn.execSql(createRequest('SELECT 42', 42, function () {
                pool.release(conn); //release the connect

                pool.acquire(function (err, conn) { //re-acquire the connection
                    assert(!err);

                    conn.execSql(createRequest('SELECT 42', 42, function () {

                        pool.drain(done);
                    }));
                });
            }));
        });
    });

    it('drain', function (done) {
        this.timeout(10000);

        var poolConfig = {min: 3};
        var pool = new ConnectionPool(poolConfig, connectionConfig);

        pool.acquire(function() { });

        setTimeout(function() {
            assert.equal(pool.connections.length, poolConfig.min);
            pool.drain(done);
        }, 4);
    });
});

describe('Load Test', function() {
    var statistics = require('simple-statistics');
    
    it('Memory Leak Detection - Connection Error', function(done) {
        this.timeout(60000);
        if (!global.gc)
            throw new Error('must run nodejs with --expose-gc');
        var count = 0;
        var mem = [];
        var groupCount = 20;
        var poolSize = 1000;
        var max = poolSize * groupCount;
        
        var pool = new ConnectionPool({ max: poolSize, min: poolSize, retryDelay: 1}, {
            userName: 'testLogin',
            password: 'wrongPassword',
            server: 'localhost'
        });
        
        pool.on('error', function() {
            if ((++count % poolSize) !== 0)
                return;
            
            global.gc();
            
            var heapUsedKB = Math.round(process.memoryUsage().heapUsed / 1024);
            mem.push([count, heapUsedKB]);
            // console.log(count + ': ' + heapUsedKB + 'KB');
            
            if (count === max) {
                var data = statistics.linearRegression(mem);
                //console.log(data.m);
                if (data.m >= 0.025)
                    done(new Error('Memory leak detected.'));
                else
                    done();
                
                pool.drain();
            }
        });
    });
    
    it('Memory Leak Detection - acquire() and Request', function(done) {
        this.timeout(60000);
        if (!global.gc)
            throw new Error('must run nodejs with --expose-gc');
    
        var poolConfig = {min: 67, max: 123};
        var pool = new ConnectionPool(poolConfig, {
            userName: 'test',
            password: 'test',
            server: 'dev1'
        });
    
        var clients = 1000;
        var connections = 100;
        var max = clients * connections;
        var mem = [];
    
        for (var i = 0; i < clients; i++)
            createClient();
    
        var count = 0;
    
        function end(err) {
            done(err);
            pool.drain();
        }
    
        function createClient() {
            var clientCount = 0;
        
            function createRequest() {
                pool.acquire(function(err, connection) {
                    if (err)
                        return end(err);
                
                    var request = new Request('select 42', function (err) {
                        if (err)
                            return end(err);
                    
                        connection.release();
                    
                        if (++clientCount < connections)
                            createRequest();
                    
                        if ((++count % 1000) === 0) {
                            global.gc();
                        
                            if (count === max) {
                                var data = statistics.linearRegression(mem);
                                //console.log(data.m);
                                if (data.m >= 0.025)
                                    end(new Error('Memory leak detected.'));
                                else
                                    end();
                            } else {
                                var heapUsedKB = Math.round(process.memoryUsage().heapUsed / 1024);
                                mem.push([count, heapUsedKB]);
                                // console.log(count + ': ' + heapUsedKB + 'KB');
                            }
                        }
                    });
                
                    connection.execSql(request);
                });
            }
        
            createRequest();
        }
    });
});


index.js
var Connection = require('tedious').Connection;
var util = require('util');

var connectionEventNames = [
  'connect',
  'end',
  'debug',
  'infoMessage',
  'errorMessage',
  'databaseChange',
  'languageChange',
  'charsetChange',
  'secure'
];

function PooledConnection(pool, config) {
    Connection.call(this, config);
    this.pool = pool;
}

util.inherits(PooledConnection, Connection);

PooledConnection.prototype.close = function() {
    this.emit('end');
    this.pool.returnConnectionToPool(this);
}

PooledConnection.prototype._close = function() {
    Connection.prototype.close.call(this);
}

function ConnectionPool(poolConfig, connectionConfig) {
  var pool = this;

  pool.config = poolConfig;
  pool.requestNumber = 0;
  pool.stats = {
    maxSize: poolConfig.maxSize,
    connections: 0,
    connectionsInUse: 0
  };
  pool.inUseConnections = [];
  pool.availableConnections = [];
  pool.pendingConnectionRequests = [];

  pool.returnConnectionToPool = function(connection) {
    connectionEventNames.forEach(function removeAllListeners(eventName) {
      connection.removeAllListeners(eventName);
    });

    pool.inUseConnections.splice(pool.inUseConnections.indexOf(connection), 1);
    pool.stats.connectionsInUse--;

    if (pool.pendingConnectionRequests.length > 0) {
      var pendingConnectionRequest = pool.pendingConnectionRequests.shift();

      pendingConnectionRequest(connection);
    }
  };

  pool.requestConnection = function(callback) {
    var requestNumber = ++pool.requestNumber;
    var connection;

    if (pool.availableConnections.length > 0) {
      connection = availableConnections.shift();
      connection.emit('connect');
    } else if (pool.inUseConnections.length < pool.config.maxSize) {
      connection = new PooledConnection(pool, connectionConfig);
      connection.number = ++pool.stats.connections;
    }

    if (connection) {
      useConnection(connection);
    } else {
      pool.pendingConnectionRequests.push(function (connection) {
        useConnection(connection);
        connection.emit('connect');
      });
    }

    function useConnection(connection) {
      pool.inUseConnections.push(connection);
      pool.stats.connectionsInUse++;

      callback(connection);
    }
  };

  return {
    requestConnection: pool.requestConnection,
    stats: pool.stats
  };
};

module.exports = ConnectionPool;


lib/connection-pool.js
'use strict';
const Connection = require('tedious').Connection;
const EventEmitter = require('events').EventEmitter;
const util = require('util');

function ConnectionPool(poolConfig, connectionConfig) {
    this.pending = new Set();
    this.free = new Set();
    this.used = new Set();
    
    this.waiting = []; //acquire() callbacks that are waiting for a connection to come available
    this.connectionConfig = connectionConfig;

    this.max = poolConfig.max >= 1 ? poolConfig.max : 50;

    this.min = Math.min(this.max, poolConfig.min >= 0 ? poolConfig.min : 10);

    this.idleTimeout = !poolConfig.idleTimeout && poolConfig.idleTimeout !== false
        ? 300000 //5 min
        : poolConfig.idleTimeout;

    this.retryDelay = !poolConfig.retryDelay && poolConfig.retryDelay !== false
        ? 5000
        : poolConfig.retryDelay;

    this.acquireTimeout = !poolConfig.acquireTimeout && poolConfig.acquireTimeout !== false
        ? 60000 //1 min
        : poolConfig.acquireTimeout;

    if (poolConfig.log) {
        if (Object.prototype.toString.call(poolConfig.log) === '[object Function]') {
            this.log = poolConfig.log;
        } else {
            this.log = function(text) {
                console.log('Tedious-Connection-Pool: ' + text);
            };
        }
    } else {
        this.log = function() {};
    }

    this.drained = false;
    
    this._nextId = 0;
    this._retryTimeout = null;
    
    setImmediate(fill.bind(this));
}

util.inherits(ConnectionPool, EventEmitter);
    
function createConnection(pooled) {
    this.log(pooled.id + ': creating connection');
    
    const connection = new Connection(this.connectionConfig);
    pooled.con = connection;
    connection.release = release.bind(this, pooled);
    
    const removeFromPool = () => {
        if (pooled.timeout)
            clearTimeout(pooled.timeout);
        
        //further indicate that this pooled object is dead to us, so it's easy to ignore
        // for example, if the user tries to release a connection that not a part of the pool
        delete pooled.con;
        
        const inPool = this.pending.delete(pooled) || this.free.delete(pooled) || this.used.delete(pooled);
        if (!inPool)
            return;
    
        if (this.drained) {//pool has been drained
            if (this.pending.size === 0 && this.free.size === 0 && this.used.size === 0)
                this.drained(); //once empty execute drained callback
            return;
        }
        
        if (!this.drained) {
            if (!this._retryTimeout) {
                this._retryTimeout = setTimeout(() => {
                    this._retryTimeout = null;
                    fill.call(this);
                }, this.retryDelay);
            }
        }
    };
    
    const handleError = (err) => {
        this.log(pooled.id + ': closing because of error');
        this.emit('error', err);
    
        removeFromPool();
    };
    
    connection.on('connect', (err) => {
        this.log(pooled.id + ': connected');
        
        if (this.drained) { //pool has been drained
            this.log(pooled.id + ': closing because pool has been drained');
            connection.close();
            return;
        }

        if (err) {
            handleError(err);
            return;
        }
        
        setFree.call(this, pooled);
    });

    connection.on('error', handleError);
    
    const endHandler = () => {
        this.log(pooled.id + ': ended');
    
        removeFromPool();
    };
    
    connection.on('end', endHandler);
}

function fill() {
    if (this.drained || this._retryTimeout) //pool has been drained
        return;
    
    const total = this.pending.size + this.free.size + this.used.size;
    
    let amount = Math.min(
        this.max - total, //max that can be created
        this.waiting.length - this.free.size); //how many are needed, minus how many are available
    
    amount = Math.max(
        this.min - total, //amount to create to reach min
        amount);
    
    if (amount > 0)
        this.log('pool: filling pool with ' + amount);
    
    for (let i = 0; i < amount; i++) {
        const pooled = {
            id: ++this._nextId
        };
        
        this.pending.add(pooled);
        
        setImmediate(createConnection.bind(this, pooled)); //avoid blocking if a lot of connections are being created
    }
}

function getOne(set) {
    for (const item in set)
        return item;
    
    return undefined;
}

ConnectionPool.prototype.acquire = function (callback) {
    if (this.drained) //pool has been drained
        return;

    const waiter = {
        callback: callback
    };

    if (this.free.size > 0) {
        const pooled = getOne(this.free);
        this.free.remove(pooled);
        setUsed.call(this, pooled, waiter);
        
    } else {
        if (this.acquireTimeout) {
            waiter.timeout = setTimeout(() => {
                for (let i = this.waiting.length - 1; i >= 0; i--) {
                    const waiter2 = this.waiting[i];

                    if (waiter2.timeout === waiter.timeout) {
                        this.waiting.splice(i, 1);
                        waiter.callback(new Error('Acquire Timeout Exceeded'));
                        return;
                    }
                }
            }, this.acquireTimeout);
        }

        this.waiting.push(waiter);
        fill.call(this);
    }
};

ConnectionPool.prototype.release = function (connection) {
    for (const pool in this.used) {
        if (pool.con === connection) {
            connection.release();
            return;
        }
    }
};

function release(pooled) {
    if (this.drained) //pool has been drained
        return;
    
    if (pooled.con) { //make sure connection is in the pool
        
        //reset connection before reusing it
        pooled.con.reset((err) => {
            if (!pooled.con) //the connection error-ed during the reset
                return;
            
            if (err) { //there is an error, don't reuse the connection, just close it
                pooled.con.close();
                return;
            }
            
            this.log(pooled.id + ': released');
            
            setFree.call(this, pooled);
        });
    }
}

function setUsed(pooled, waiter) {
    this.used.add(pooled);
    
    if (pooled.timeout) {
        clearTimeout(pooled.timeout);
        pooled.timeout = undefined;
    }
    if (waiter.timeout) {
        clearTimeout(waiter.timeout);
        waiter.timeout = undefined;
    }
    this.log(pooled.id + ': acquired');
    setImmediate(() => {
        waiter.callback(null, pooled.con);
    });
}

function setFree(pooled) {
    this.pending.delete(pooled);
    this.used.delete(pooled);
    
    //check if there is anything waiting for a connection
    const waiter = this.waiting.shift();
    
    if (waiter !== undefined) {
        setUsed.call(this, pooled, waiter);
        return;
    }
    
    this.free.add(pooled);

    //start idle timeout for pooled connection
    pooled.timeout = setTimeout(() => {
        this.log(pooled.id + ': closing idle connection');
        pooled.con.close();
    }, this.idleTimeout);
}

ConnectionPool.prototype.drain = function (callback) {
    if (this.drained) //pool has been drained
        throw new Error('pool: already drained');
    
    this.log('pool: draining');
    
    this.drained = callback || function() {};

    for (let i = this.waiting.length - 1; i >= 0; i--) {
        const waiter = this.waiting[i];

        if (waiter.timeout)
            clearTimeout(waiter.timeout);
    }

    this.waiting = null;
    
    if (this.pending.size === 0 && this.free.size === 0 && this.used.size === 0) {
        callback();
    } else {
        for (const pooled of this.pending) {
            if (pooled.con)
                pooled.con.close();
        }
    
        for (const pooled of this.free)
            pooled.con.close();
    
        for (const pooled of this.used)
            pooled.con.close();
    }
    
    this.pending = null;
    this.free = null;
    this.used = null;
};

module.exports = ConnectionPool;
